### ISSUE TITLE
Tuple.wrap(Collections$UnmodifiableRandomAccessList)  can't work!!!

### ISSUE THREAD

**Issue by Bleoo** [eclipse-vertx/vertx-sql-client#801]
### Questions

I am using pgclient.
Because my sql parameter was placed in `bindValues`, it's an `Collections$UnmodifiableRandomAccessList`, when I used it in this way.

```
client.preparedQuery(sql).execute(Tuple.wrap(bindValues))
```

I found that it not be executed normally, and no error was reported, but that `promise` disappeared !!! It cannot execute to the onComplete method !!!

But in version 3.8 I can use it like this.

Later, I can only use it in the other way and it works normally. it's kotlin code.

```
client.preparedQuery(sql).execute(Tuple.wrap(*bindValues.toTypedArray()))

```
or this:
```
client.preparedQuery(sql).execute(Tuple.tuple(bindValues))
```
If so, I have to replace all the code that uses this method. i can'nt do that.

Here is the difference between the two tuples. tuple1 can work, tuple2 cann't.
![image](https://user-images.githubusercontent.com/11569795/98798875-57d93f80-2449-11eb-9ca7-09963b120cbc.png)

pls help ~

### Version

4.0.0.CR1

### Extra

jdk8

**Comment by Bleoo** [comment]
@vietj

**Comment by vietj** [comment]
can you provide a reproducer ?

**Comment by vietj** [comment]
@Bleoo a reproducer case ?

**Comment by mocukie** [comment]
Same question here.

A reproducer:
```java
Future.succeededFuture(pgPool)
		.compose(client ->
			client.preparedQuery("select 0 where 1 = $1")
				.execute(Tuple.wrap(Stream.of(1).collect(Collectors.toUnmodifiableList())))
		)
		.onSuccess(r -> System.out.println("success"))
		.onFailure(e -> {
			System.out.println("failed");
			e.printStackTrace();
		});
```
The exeception was occurred in `PgParamDesc#prepare`, `values.setValue(i, val)` ,  but`CommandWaiter` in `Poolbase#schedule` is not override method handleException.
If change `Future.succeededFuture(pgPool)` to `pgPool.getConnection()`, will print the UnsupportedOperationException.

**Version**
4.0.0.CR2
java 11

**Comment by BillyYccc** [comment]
A simple reproducer would be 
```
  @Test
  public void reproducer() {
    List<Integer> list = new ArrayList<>();
    list.add(1);
    list.add(2);
    list.add(3);
    Tuple tuple = Tuple.wrap(list);
    Assert.assertEquals(3, tuple.size());
  }
```

I think it's because the `Tuple#wrap` has two overload methods `Tuple#wrap(List<Object> list)` and `Tuple#wrap(Object... array)` while the former one accepts a generic List<Object> and the actual parameter in use might be `List<Integer>`, `List<Long>` or whatever else, and Java generics are invariant so it will construct an incorrect Tuple of list of list using the latter factory method.

**Comment by vietj** [comment]
thanks @BillyYccc

**Comment by mocukie** [comment]
Thanks, but java type inference seems to choose `Tuple#wrap(List<Object> list)`, because I did not to use a type witness on `Collectors.toUnmodifiableList()`.
If change `Collectors.toUnmodifiableList()` to `Collectors.toList()` that would work.

**Comment by vietj** [comment]
java s*cks big times

**Comment by vietj** [comment]
we should likely rename these wrapList / wrapArray and remove wrap . and deprecate this in 3.9

**Comment by vietj** [comment]
the workaround is in user code you need to write `Tuple.wrap((List)list)` to force using the correct version. We could work around by check whether it's an array of 1 element and the first element is a List but well it's convoluted

**Comment by vietj** [comment]
I believe the obvious change is to use instead:

```java
  static <T> Tuple wrap(List<T> list) {
    return new ListTuple((List<Object>) list);
  }

  static Tuple wrap(Object[] array) {
    return new ListTuple(Arrays.asList(array));
  }
```

**Comment by BillyYccc** [comment]
Another concern is that you should avoid wrapping immutable list as `Tuple` params because the parameters might get mutated at runtime.

**Comment by vietj** [comment]
the tuple by vertx are indeed modified, perhaps we should internally detect when it's a foreign tuple and copy the underlying list on a write

**Comment by Bleoo** [comment]
It has changed the original way of use, so how to solve this problem in the end?

**Comment by vietj** [comment]
it's not clear what you are meaning for @Bleoo, the methods got the signature changed so that calling Tuple.wrap(list) will call Tuple.wrap(List) and not Tuple.wrap(Object[])

**Comment by vietj** [comment]
can you provide a reproducer for your case @Bleoo, others provided a reproducer using Tuple.wrap but I can see your code does not use it, however you never provided a clear reproducer for your case

**Comment by Bleoo** [comment]
Thank you for your reply.
I'm sorry to be busy these days, so I didn't have time to reply.
It is my reproducer (kotlin) : 
``` kotlin
val unmodifiableList = Collections.unmodifiableList(listOf(1)).toList()
  pgPool.preparedQuery("select 0 where 1 = $1")
    .execute(Tuple.wrap(unmodifiableList))
    .onSuccess { r -> println("success") }
    .onFailure { e ->
      println("failed, but can't println this")
      e.printStackTrace()
    }
```
My case is an unmodifiable list, which is generated by JOOQ through SQL. So, I don't want to change this usage after vertx upgrading.

**Comment by vietj** [comment]
in this case you should not call wrap, wrap will just wrap the list, I think instead we can add a Tuple.tuple(List) to accomodate this case.

**Comment by vietj** [comment]
@Bleoo see https://github.com/eclipse-vertx/vertx-sql-client/issues/845

**Comment by Bleoo** [comment]
but in version 3.8 I can use it like this, Tuple.wrap(unmodifiable lists) .
I'm confused. Isn't the version down compatible? @vietj

**Comment by vietj** [comment]
this will be backported to 3.9.5 release

for now this will be available in 4.0.0
